File: node-root/snake-ai/scripts/performanceTest/multiThread.ts
import * as fs from "fs";
import * as path from "path";
import SnakeBrain from "../../SnakeBrain";
import MultiThreadGames from "../../MultiThreadGames";
import CalcUtils from "../../CalcUtils";
import { TimingUtils } from "./Timer";
import type { ISnakeBrain } from "../../SnakeBrain";

async function main() {
  const brainDataPath = path.join(__dirname, "trained-brain.json");
  const brainData: ISnakeBrain = JSON.parse(fs.readFileSync(brainDataPath, "utf8"));

  const snakeBrain = new SnakeBrain({
    inputLength: brainData.inputLength,
    layerShapes: brainData.layerShapes,
    hiddenLayerActivationFunction: brainData.hiddenLayerActivationFunction,
    providedWeightsAndBiases: {
      weights: brainData.weights,
      biases: brainData.biases,
    },
  });

  const multiThreadGames = new MultiThreadGames(10);

  const games = 10000;
  const brains = Array.from({ length: games }, () => snakeBrain);
  let score: number[] = [];

  const execTime = await TimingUtils.execTime(async () => {
    const resultPromise = brains.map((brain) => multiThreadGames.playGames(20, 20, 1, brain.toPlainObject()));
    const workerResults = await Promise.all(resultPromise);
    score = workerResults.map((x) => CalcUtils.meanOfArray(x.snakeLengthArr));
  });

  const totalTime = execTime;
  console.log(`Total time: ${totalTime.toFixed(3)}s`);

  const bestScore = Math.max(...score);
  console.log(`Best score: ${bestScore}`);

  const avgScore = CalcUtils.meanOfArray(score);
  console.log(`Average score: ${avgScore.toFixed(2)}`);

  multiThreadGames.destroy();
}

main();

File: kotlin-root/gasnake/src/main/kotlin/snake/ai/performance/MultiThread.kt
package com.example.snake.ai.performance


import com.example.snake.ai.ProvidedWeightsAndBiases
import com.example.snake.ai.SnakeBrain
import com.example.snake.ai.SnakeBrainData
import com.google.gson.Gson
import snake.ai.multithread.MultiThreadGames
import java.io.InputStreamReader

fun main() {
    val inputStream = Thread.currentThread().contextClassLoader.getResourceAsStream("trained-brain.json")
        ?: throw IllegalArgumentException("Resource not found")
    val reader = InputStreamReader(inputStream)
    val brainData: SnakeBrainData = Gson().fromJson(reader, SnakeBrainData::class.java)
    reader.close()

    val snakeBrain = SnakeBrain(
        com.example.snake.ai.Options(
            brainData.inputLength,
            brainData.layerShapes,
            brainData.hiddenLayerActivationFunction,
            ProvidedWeightsAndBiases(
                brainData.weights,
                brainData.biases
            )
        )
    )

    val multiThreadGames = MultiThreadGames(10)

    val games = 10000
    val snakeBrains = List(games) { snakeBrain }
    val scores = mutableListOf<Int>()

    val execTime = TimingUtils.execTime {
        val resultFutures = snakeBrains.map { multiThreadGames.playGames(20, 20, 1, it.toPlainObject()) }
        val workerResults = resultFutures.map { it.get() }
        scores.addAll(workerResults.map { it.snakeLengthArr.average().toInt() })
    }

    val totalTime = execTime
    println("Total time: ${"%.3f".format(totalTime)}s")

    val bestScore = scores.maxOrNull() ?: 0
    println("Best score: $bestScore")

    val avgScore = scores.average()
    println("Average score: ${"%.2f".format(avgScore)}")

    multiThreadGames.destroy()
}